### PR TITLE
Add SeaSolver owners proposer

### PR DESCRIPTION
**Pull request by sunce86**
Fixes https://github.com/cowprotocol/services/issues/715

### Test Plan

Added e2e test.

### CODE REVIEW

File: crates/shared/src/bad_token/token_owner_finder.rs [hunk 1]
```diff
@@ -1,6 +1,7 @@
 pub mod blockscout;
 pub mod ethplorer;
 pub mod liquidity;
+pub mod seasolver;
```
Comment (nlordell) [inline crates/shared/src/bad_token/token_owner_finder.rs:4]: nit: Again, I wouldn't call this `seasolver`. While they are the ones with exotic tokens and want to provide a "token owners API", we should name it more generically, as potentially other solvers may want to use the same interface for providing token owner addresses for exotic tokens.

File: crates/shared/src/bad_token/token_owner_finder.rs [hunk 2]
```diff
@@ -84,6 +88,14 @@ pub struct Arguments {
         default_value = "",
     )]
     pub whitelisted_owners: HashMap<H160, Vec<H160>>,
+
+    /// The SeaSolver url to query the token owner pairs.
+    #[clap(long, env, default_value = "https://seasolver.dev/token_holders")]
+    pub seasolver_token_owners_url: Url,
```
Comment (nlordell) [inline crates/shared/src/bad_token/token_owner_finder.rs:94]: I think this should be a `Vec<Url>` as it is very imaginable that we want to have it so other external solver teams also provide token owner lists for their exotic tokens.

Also, I don't think we should provide a default URL, or enable it by default.

File: crates/shared/src/bad_token/token_owner_finder.rs [hunk 3]
```diff
@@ -120,13 +132,21 @@ pub enum TokenOwnerFindingStrategy {
...
 impl TokenOwnerFindingStrategy {
     /// Returns the default set of token owner finding strategies.
     pub fn defaults_for_chain(chain_id: u64) -> &'static [Self] {
         match chain_id {
-            1 => &[Self::Liquidity, Self::Blockscout, Self::Ethplorer],
+            1 => &[
+                Self::Liquidity,
+                Self::Blockscout,
+                Self::Ethplorer,
+                Self::SeaSolver,
```
Comment (nlordell) [inline crates/shared/src/bad_token/token_owner_finder.rs:148]: I would not enable this by default in code. It seems very specific to that external solver, and, in theory, we only would want to enable it when running with that external solver.

File: crates/shared/src/bad_token/token_owner_finder/seasolver.rs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+pub struct SeaSolverConfiguration {
+    pub url: Url,
+    pub client: Client,
+    pub update_interval: Duration,
+}
+
+impl SeaSolverConfiguration {
+    /// Return type is `Token, Option<Owner>` because there are
+    /// entries containing `Null` instead of owner address.
+    async fn query(&self) -> Result<HashMap<Token, Option<Owner>>> {
```
Comment (nlordell) [inline crates/shared/src/bad_token/token_owner_finder/seasolver.rs:26]: Would the API scale better if they had something like:
```
https://seasolver.dev/token_holders?token=0x....
```

**Review by nlordell — COMMENTED** [review]
Nice! I do think this should be implemented in a more generic way though so that other solvers can also provide token holder addresses for exotic tokens they want to support.

**Review by MartinquaXD — COMMENTED** [review]
Mostly very minor nits, but I agree with Nick that this is a fairly generic implementation and therefore doesn't need to have the `seasolver` name.

File: crates/shared/src/bad_token/token_owner_finder.rs (hunk 2 above)
Comment (sunce86) [reply crates/shared/src/bad_token/token_owner_finder.rs]: Changed to `Vec<Url>`. I agree this is a good suggestion only if other solvers will not require any additional argument, like api-key or something similar.

File: crates/shared/src/bad_token/token_owner_finder/seasolver.rs (hunk 4 above)
Comment (sunce86) [reply crates/shared/src/bad_token/token_owner_finder/seasolver.rs]: Might be, but I guess the caching would be more complex to implement than... Not sure if worth it

File: crates/shared/src/bad_token/token_owner_finder/seasolver.rs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    #[ignore]
+    async fn seasolver_e2e_test() {
+        let url = std::env::var("SEASOLVER_TOKEN_HOLDERS").unwrap();
+        let configuration = SeaSolverConfiguration {
+            url: Url::from_str(&url).unwrap(),
+            client: Client::new(),
+            update_interval: Duration::from_secs(1000),
+        };
+        let finder = AutoUpdatingSeaSolverTokenOwnerFinder::new(configuration);
+        tokio::time::sleep(Duration::from_secs(10)).await;
```
[the remark this answers is not in the document]
Comment (sunce86) [reply crates/shared/src/bad_token/token_owner_finder/seasolver.rs]: Probably. Did not want to waste time tweaking it :)

File: crates/shared/src/bad_token/token_owner_finder/seasolver.rs [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+#[cfg(test)]
+mod tests {
+    use std::str::FromStr;
+
+    use super::*;
+
+    #[tokio::test]
+    #[ignore]
+    async fn seasolver_e2e_test() {
+        let url = std::env::var("SEASOLVER_TOKEN_HOLDERS").unwrap();
```
[the remark this answers is not in the document]
Comment (sunce86) [reply crates/shared/src/bad_token/token_owner_finder/seasolver.rs]: Did not want to leak it yes. Removed the value from arguments also so it's not visible anymore.

File: crates/shared/src/bad_token/token_owner_finder.rs [hunk 7]
```diff
@@ -84,6 +90,15 @@ pub struct Arguments {
...
+
+    /// The solvers urls to query the token owner pairs.
+    #[clap(long, env, use_value_delimiter = true)]
+    pub solver_token_owners_urls: Vec<Url>,
+
+    /// Interval between consecutive queries to update the solver token owner pairs.
+    /// Values should be in pair with `solver_token_owners_urls`
+    #[clap(long, env, use_value_delimiter = true, value_parser = duration_from_seconds)]
+    pub solver_token_owners_cache_update_intervals: Vec<Duration>,
 }
```
Comment (sunce86) [inline crates/shared/src/bad_token/token_owner_finder.rs:102]: Not sure if we should force these two values to be paired.

**Comment by sunce86** [comment]
Did a refactor by introducing the `TokenOwnerSolverApi` for a solution to be more generic and future proof.
This way we can reuse the caching mechanism for all solvers while inner communication with solver api is implemented individually for each one.
One thing I dislike is the arguments definition. If next solver requires api key or something, will need to change the handling of parameters.

File: crates/shared/src/bad_token/token_owner_finder.rs (hunk 7 above)
Comment (nlordell) [reply crates/shared/src/bad_token/token_owner_finder.rs:102]: We can probably start with defining one value for all `solver_token_owners_urls`, but its fine like this too.

**Review by nlordell — COMMENTED** [review]
Seems good to me.

I would have, intuitively, thought it made more sense for the API to be per-token instead of being "give me all the token owners for all tokens". Is there a reason we didn't go this way (besides requiring the SeaSolver team to change their endpoint's behaviour?).

**Comment by sunce86** [comment]
> Is there a reason we didn't go this way

The most important reason is that we get all the token/owner pairs cached with 1 http request. Any further calls to get the candidates would be fetched from the cache therefore avoiding the unnecessary http requests.

File: crates/shared/src/bad_token/token_owner_finder.rs [hunk 8]
```diff
@@ -84,6 +88,14 @@ pub struct Arguments {
...
     )]
     pub whitelisted_owners: HashMap<H160, Vec<H160>>,
+
+    /// The SeaSolver url to query the token owner pairs.
+    #[clap(long, env, default_value = "https://seasolver.dev/token_holders")]
+    pub seasolver_token_owners_url: Url,
+
+    /// Interval between consecutive queries to update the SeaSolver token owner pairs.
+    #[clap(long, env, default_value = "300", value_parser = duration_from_seconds)]
+    pub seasolver_token_owners_cache_update_interval: Duration,
```
[the remark this answers is not in the document]
Comment (sunce86) [reply crates/shared/src/bad_token/token_owner_finder.rs]: Added comment.

**Comment by nlordell** [comment]
> Any further calls to get the candidates would be fetched from the cache therefore avoiding the unnecessary http requests.

Right, so the question becomes do we prefer a single large HTTP request upfront, or multiple small ones. I agree that if the list of owners is small, then this approach makes more sense.

**Review by vkgnosis — APPROVED** [review]
Code looks fine to me. There is some bike shedding about the api. To me the most correct api would provide exactly what TokenOwnerProposing needs which is

    /// Find candidate addresses that might own the token.
    async fn find_candidate_owners(&self, token: H160) -> Result<Vec<H160>>;

Then we can call these like any other source during the bad token check. How someone implementing this api does on their backend doesn't matter to us (whether they store a hashmap, how they cache).

I'm ok with the current state of the PR too. That has the advantage of making it slightly easier to implement but the downside of not fitting what we need perfectly. I somewhat prefer the more correct api unless there is a good reason to not do it this way.

**Review by nlordell — APPROVED** [review]
Sorry, didn't mean to bikeshed and meant to approve with my previous comment.

In general I agree with @vkgnosis that the API should probably more closely model our internal API (i.e. `GET /token-owners?token=0x...`), but I also think this approach is fine for how we plan on using it now.

**Pull request merged**